File: src/main/java/com/indiepost/controller/api/admin/AdminTagController.java
package com.indiepost.controller.api.admin;

import com.indiepost.dto.TagDto;
import com.indiepost.service.TagService;
import org.springframework.web.bind.annotation.*;

import javax.inject.Inject;
import java.util.List;

@RestController
@RequestMapping(value = "/api/admin/tag", produces = {"application/json; charset=UTF-8"})
public class AdminTagController {

    private final TagService tagService;

    @Inject
    public AdminTagController(TagService tagService) {
        this.tagService = tagService;
    }

    @PostMapping
    public TagDto createNewTag(@RequestBody TagDto tagDto) {
        return tagService.save(tagDto);
    }

    @GetMapping
    public List<TagDto> searchTags(@RequestParam String tagName) {
        return tagService.findByNameLike(tagName);
    }
}

File: src/main/kotlin/com/indiepost/controller/admin/AdminTagController.kt
package com.indiepost.controller.admin

import com.indiepost.dto.SelectedTagDto
import com.indiepost.dto.TagDto
import com.indiepost.service.TagService
import org.slf4j.LoggerFactory
import org.springframework.cache.annotation.CacheEvict
import org.springframework.data.domain.Pageable
import org.springframework.web.bind.annotation.*
import javax.inject.Inject

@RestController
@RequestMapping("/admin/tags")
class AdminTagController @Inject constructor(private val tagService: TagService) {

    @GetMapping
    fun getTags(@RequestParam pageable: Pageable): List<TagDto> {
        return tagService.find(page = pageable.pageSize, size = pageable.pageSize)
    }

    @PostMapping("/selected")
    @CacheEvict(cacheNames = ["tag::rendered", "post::rendered", "home::rendered", "static-page::rendered"], allEntries = true)
    fun updateSelectedTags(@RequestBody tags: List<String>) {
        logger.info("Clear all caches")
        tagService.updateSelected(tags)
    }

    @GetMapping("/selected")
    fun getSelectedTags(): List<SelectedTagDto> {
        return tagService.findSelected()
    }

    companion object {
        private val logger = LoggerFactory.getLogger(AdminTagController::class.java)
    }
}
